students/chieu_quach/lesson07/html_render.py

```diff
@@ -0,0 +1,353 @@
+            # out_file.write("<{}>\n".format(self.tag))
+                out_file.write("<{}>".format(self.tag))
+            for content in self.contents:
+           #    print("content u", content)
+                try:                
+                    content.render(out_file)
+                   
+                except AttributeError:
+                    if content == "DOCTYPE":
+                        pass
+                    elif self.tag == "meta":
+                        out_file.write(indentation(content, head_indent_txt))
+                    elif self.tag == "pz":
+                        out_file.write(indentation(stylename_p, head_indent_txt))
+                        out_file.write(indentation(content, p_indent_txt))
+                    elif self.tag == "ul":
+                        out_file.write(indentation(stylename_ul, head_indent_txt))
+                    elif self.tag == "li":
+                        out_file.write(indentation(content, indent_txt))
+                    elif self.tag == "liu":
+                        out_file.write(indentation(stylename_li, p_indent_txt))
+                        out_file.write(indentation(content, indent_txt))
+                    else:
+                        out_file.write(content)
+                
+                    # if tag == title write output in one line including <title>
+                    if self.tag != "title":
+                        out_file.write("\n")
+                    
+
+           #   print ("self.tag ", self.tag)      
+           # prints the output content message             
+           #   print ("self.contents ", self.contents)
+            if self.tag == "liu":
+                self.tag = "li"
+            if self.tag == "pz":
+                self.tag = "p"
+
+            nameformat = "</{}>\n".format(self.tag)
+            if self.tag == "hr":
+          
+                out_file.write(indentation(nameformat, head_indent_txt))  
+           # out_file.write("<{} />\n".format(self.tag))
+           #   out_file.write(indentation(nameformat, head_indent_txt))
+            elif self.tag == "meta":
+                pass
+            elif self.tag == "head":
+         #   nameformat = "</{}>\n".format(self.tag)
+                out_file.write(indentation(nameformat, head_indent))
+            elif self.tag == "p":
+        #    nameformat = "</{}>\n".format(self.tag)
+                out_file.write(indentation(nameformat, head_indent_txt))
+            elif self.tag == "li":
+                out_file.write(indentation(nameformat, p_indent_txt))
+            elif self.tag == "ul":
+                out_file.write(indentation(nameformat, head_indent_txt))
+            elif self.tag == "body":
+                out_file.write(indentation(nameformat, head_indent))
+            else:
+            
+                out_file.write("</{}>\n".format(self.tag))
+
+
+
+            
+#     #   out_file.write("just something as a place holder...")
```

Comment:
this is what the solution should look close to:
```

class Element:

    tag = "html"
    indent = "    "

    def __init__(self, content=None, **kwargs):
        self.attributes = kwargs
        self.content = []
        if content:
            # call the classes append method
            # so that it can do anything special it needs to do
            self.append(content)

    def append(self, content):
        """
        add a new piece of content or another element to this element
        """
        # note: this changed the internal representation of content
        #       it no longer holds strings -- so a test will fail
        #       but that test was testing internal API --
        #       it's probably better remove it
        if hasattr(content, 'render'):
            self.content.append(content)
        else:
            self.content.append(TextWrapper(str(content)))

    def make_tags(self):
        """
        create the tags
        -- in a separate method so different subclass' render methods can use it
        """
        attrs = " ".join(['{}="{}"'.format(key, val) for key, val in self.attributes.items()])
        if attrs.strip():
            open_tag = "<{} {}>".format(self.tag, attrs.strip())
        else:
            open_tag = "<{}>".format(self.tag)
        close_tag = "</{}>".format(self.tag)

        return open_tag, close_tag

    def render(self, out_file, cur_ind=""):
        open_tag, close_tag = self.make_tags()
        out_file.write(cur_ind + open_tag + "\n")
        for stuff in self.content:
            stuff.render(out_file, cur_ind + self.indent)
            out_file.write("\n")
        out_file.write(cur_ind + close_tag)
```


I believe the issue with your code appear to be due to not utilizing the inheritance